lib/SIL/SILConstants.cpp

```diff
@@ -0,0 +1,405 @@
+  assert(getKind() == Unknown);
+  return value.unknown->reason;
+}
+
+//===----------------------------------------------------------------------===//
+// Higher level code
+//===----------------------------------------------------------------------===//
+
+/// The SIL location for operations we process are usually deep in the bowels
+/// of inlined code from opaque libraries, which are all implementation details
+/// to the user.  As such, walk the inlining location of the specified node to
+/// return the first location *outside* opaque libraries.
+static SILDebugLocation skipInternalLocations(SILDebugLocation loc) {
+  auto ds = loc.getScope();
+
+  if (!ds || loc.getLocation().getSourceLoc().isValid())
+    return loc;
+
+  // Zip through inlined call site information that came from the
+  // implementation guts of the library.  We want to report the message inside
+  // the user's code, not in the guts we inlined through.
+  for (; auto ics = ds->InlinedCallSite; ds = ics) {
+    // If we found a valid inlined-into location, then we are good.
+    if (ds->Loc.getSourceLoc().isValid())
+      return SILDebugLocation(ds->Loc, ds);
+    if (SILFunction *F = ds->getInlinedFunction()) {
+      if (F->getLocation().getSourceLoc().isValid())
+        break;
+    }
+  }
+
+  if (ds->Loc.getSourceLoc().isValid())
+    return SILDebugLocation(ds->Loc, ds);
+
+  return loc;
+}
+
+/// Dig through single element aggregates, return the ultimate thing inside of
+/// it.  This is useful when dealing with integers and floats, because they
+/// are often wrapped in single-element struct wrappers.
+SymbolicValue SymbolicValue::lookThroughSingleElementAggregates() const {
+  auto result = *this;
+  while (1) {
+    if (result.getKind() != Aggregate)
+      return result;
+    auto elts = result.getAggregateValue();
+    if (elts.size() != 1)
+      return result;
+    result = elts[0];
+  }
+}
+
+/// Emits an explanatory note if there is useful information to note or if there
+/// is an interesting SourceLoc to point at.
+/// Returns true if a diagnostic was emitted.
+static bool emitNoteDiagnostic(SILInstruction *badInst, UnknownReason reason,
+                               SILLocation fallbackLoc, std::string error) {
+  auto loc = skipInternalLocations(badInst->getDebugLocation()).getLocation();
+  if (loc.isNull()) {
+    // If we have important clarifying information, make sure to emit it.
+    if (reason == UnknownReason::Default || fallbackLoc.isNull())
+      return false;
+    loc = fallbackLoc;
+  }
+
+  auto &module = badInst->getModule();
+  diagnose(module.getASTContext(), loc.getSourceLoc(),
+           diag::constexpr_unknown_reason, error)
+      .highlight(loc.getSourceRange());
+  return true;
+}
+
+/// Given that this is an 'Unknown' value, emit diagnostic notes providing
+/// context about what the problem is.
+void SymbolicValue::emitUnknownDiagnosticNotes(SILLocation fallbackLoc) {
```

Comment:
I don't want to move this logic into a pass because there may be multiple different passes that want to use it to emit diagnostics about unknown values. (e.g. some of the tensorflow passes need to const-eval certain things and emit diagnostics when the const-eval fails).

Having it in SymbolicValue seems like the easiest way for passes who use it to call it, and I don't see much harm in leaving it here. Alternatively, I could turn it into a standalone utility function. Maybe `emitUnknownDiagnosticNotes(SILConstant unknown, SILLocation fallbackLoc)` in `SILOptimizer/Utils/ConstantDiagnostics`? What do you prefer?